cpp/bench/ann/src/common/benchmark.hpp

```diff
@@ -38,11 +38,59 @@
 #include <string>
 #include <unistd.h>
 #include <vector>
+
 namespace raft::bench::ann {
 
-std::mutex init_mutex;
-std::condition_variable cond_var;
-std::atomic_int processed_threads{0};
+/**
+ * A simple progress tracker that allows syncing threads multiple times and resets the global
+ * progress once the threads are done.
+ */
+struct progress_sync {
+  progress_sync() = default;
+  ~progress_sync() noexcept
+  {
+    done_.store(true);
+    cv_.notify_all();
+    auto rem = total_progress_.fetch_sub(thread_progress_);
+    if (rem == thread_progress_) {
+      // the last thread to exit clears the progress state.
+      done_.store(false);
+    }
+  }
+
+  /** Advance the progress counter by `n`. */
+  auto progress(int n)
+  {
+    thread_progress_ += n;
+    auto prev = total_progress_.fetch_add(n);
+    cv_.notify_all();
+    return prev;
+  }
+
+  /** Wait till the progress counter reaches `n` or finishes abnormally. */
+  auto wait_for_progress(int limit)
+  {
+    int cur = total_progress_.load();
+    if (cur >= limit) { return cur; }
+    auto done = done_.load();
+    if (done) { return cur; }
+    std::unique_lock lk(mutex_);
```

Comment:
In theory we can, but that would increase the time the mutex is locked and thus increase contention. I've added these two early stop conditions to leave the chance to avoid touching the mutex.